File: 0076-minimum-window-substring/0076-minimum-window-substring.js
/**
 * @param {string} s
 * @param {string} t
 * @return {string}
 */
var minWindow = function(s, t) {
    if (s.length < t.length || !t.length || !s.length) return "";

    let start = 0, end = 0, strStart = 0, minLength = t.length + s.length, count = t.length;
    const mapper = Array(128).fill(0)

    for (const c of t) mapper[c.charCodeAt(0)] += 1;
    
    while (end < s.length) {
        if (mapper[s[end].charCodeAt(0)] > 0) count -= 1;
        mapper[s[end].charCodeAt(0)] -= 1;
        end += 1;

        while (count === 0) {
            if (end - start < minLength ) {
                minLength = end - start;
                strStart = start;
            };

            if (mapper[s[start].charCodeAt(0)] === 0) count += 1;
            mapper[s[start].charCodeAt(0)] += 1;
            start += 1;
        };
    };

    if (minLength === t.length + s.length) return "";

    return s.slice(strStart, strStart + minLength);
};

File: 0076-minimum-window-substring/0076-minimum-window-substring.py
class Solution:
    def minWindow(self, s: str, t: str) -> str:
        if not s or not t or len(s) < len(t):
            return ""
        
        mapper = [0] * 128
        count = len(t)
        start = 0
        end = 0
        l = float('inf')
        
        curr_start = 0

        for char in t:
            mapper[ord(char)] += 1

        while end < len(s):
            if mapper[ord(s[end])] > 0:
                count-=1

            mapper[ord(s[end])]-=1
            end+=1

            while count == 0:
                if end - start < l:
                    curr_start = start
                    l = end - start

                if mapper[ord(s[start])] == 0:
                    count+=1
                
                mapper[ord(s[start])]+=1
                start+=1

        return "" if l == float('inf') else s[start-1:end+1]



            


